superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+        std::cout << "Error: failed to set socket." << endl;
+        MPI_Finalize();
+    }
+    return local_port;
+}
+
+// Get and broadcast ports for each run
+vector<int> prepare_ports(const vector<std::pair<int, int>> &run_pairs_in_parallel) {
+    int pair_count = run_pairs_in_parallel.size();
+    vector<int> ports(pair_count);
+
+    for (int index = 0; index < run_pairs_in_parallel.size(); index++) {
+        int server_index = run_pairs_in_parallel[index].first;
+        // server get a free port and send to rank 0
+        if (server_index == world_rank) {
+            int port = get_available_port();
+            MPI_Send(&port, 1, MPI_INT, 0, index, MPI_COMM_WORLD);
+        }
+        // rank 0 recv port from server
+        if (world_rank == 0) {
+            int port;
+            MPI_Recv(&port, 1, MPI_INT, server_index, index, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+            ports[index] = port;
+        }
+    }
+    // rank 0 broadcast ports to all ranks
+    MPI_Bcast(ports.data(), pair_count, MPI_INT, 0, MPI_COMM_WORLD);
+    return ports;
+}
+
+// Get and gather ips for all ranks
+void gather_ips(vector<string> &ips) {
+    // get my ip
+    string ip = get_my_ip();
+    // convert my ip address from string to vector<char>
+    int my_len = ip.size();
+    int max_len = 16;
+    char my_str_padded[max_len + 1];
+    memset(my_str_padded, '\0', max_len + 1);
+    memcpy(my_str_padded, ip.c_str(), my_len);
+    // prepare buffer to save ips from all ranks
+    char *all_str = NULL;
+    int all_len = (max_len + 1) * world_size;
+    all_str = (char *)malloc(all_len * sizeof(char));
+    memset(all_str, '\0', all_len);
+
+    // gather ips from all ranks
+    MPI_Allgather(my_str_padded, max_len + 1, MPI_CHAR, all_str, max_len + 1, MPI_CHAR, MPI_COMM_WORLD);
+
+    // restore all ips from char* to vector<string>
+    char *str_idx = all_str;
+    for (int rank_idx = 0; rank_idx < world_size; rank_idx++) {
+        ips.push_back(str_idx);
+        str_idx = str_idx + max_len + 1;
+    }
+}
+
+// Parse raw output of ib command
+float process_raw_output(string output) {
+    float res = -1.0;
+    try {
+        vector<string> lines;
+        split_string(output, lines, "\n");
+        regex re("\\d+\\s+\\d+\\s+\\d+\\.\\d+\\s+(\\d+\\.\\d+)\\s+\\d+\\.\\d+");
+        for (string line : lines) {
+            smatch m;
+            if (regex_search(line, m, re))
+                res = std::max(res, stof(m.str(1)));
+        }
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse raw_output: " << output << std::endl;
+    }
+
+    return res;
+}
+
+// Run ib command on server/client with server ip
+float run_cmd(string cmd_prefix, int port, bool server, string ip) {
+    string is_server = server ? "server" : "client";
+    // client sleep 1s in case that client starts before server
+    if (!server) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
This might not be enough to strictly make sure server is launched before client. One idea is to sync over MPI_Send/Recv.